### ISSUE TITLE
Bug: class labels not preserved

### ISSUE THREAD

**Issue by romanlutz** [fairlearn/fairlearn#339]
```
>>> eg = ExponentiatedGradient(constraints=DemographicParity(), estimator=LogisticRegression())
>>> eg.fit(pd.DataFrame([[1], [2], [3]]), [2, 0, 0], sensitive_features=[3,2,4])
>>> eg.predict([[1], [2], [3]])
0    0
1    0
2    1
dtype: int32
```

**Comment by romanlutz** [comment]
@MiroDudik should we restrict to just 0/1? We definitely want to consider multi-class classification in the not so distant future, so I don't think it makes sense.

**Comment by adrinjalali** [comment]
I've been wanting to mention that there's an issue with handling the outputs in the current codebase. As examples, here's how we handle the output in the `GradientBoostingClsasifier`:

https://github.com/scikit-learn/scikit-learn/blob/dcfb3df9a3df5aa2a608248316d537cd6b3643ee/sklearn/ensemble/_gb.py#L1095-L1105

and in `HistGradientBoostingClassifier`:

https://github.com/scikit-learn/scikit-learn/blob/dcfb3df9a3df5aa2a608248316d537cd6b3643ee/sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py#L1129-L1142

**Comment by romanlutz** [comment]
If this is addressed after the completion of #332 then we need to adjust the parts where we expect labels 0/1.

**Comment by adrinjalali** [comment]
generally, you just need to limit the case to a binary classification, not necessarily a `0/1` output, and where not clear, you can accept a `pos_label` to set the positive output.

**Comment by romanlutz** [comment]
I agree. At least, that applies unless we have logic that relies on 0/1 like in PR #332 
The other alternative is to store the mapping from original labels to 0...m
I'm very much thinking of the longer term here because we won't always want to restrict Exponentiated Gradient to binary classification.

**Comment by riedgar-ms** [comment]
Should this be closed? @moprescu  signed off on the fix.

**Comment by romanlutz** [comment]
Which fix?

**Comment by MiroDudik** [comment]
Our reductions for fair classification strongly rely on y being 0/1. Going to multi-class would require non-trivial algorithmic extensions. 

Handling arbitrarily encoded **binary** classification should be doable, but it is more involved than meets the eye, because:
* Fairness criteria / fairness metrics rely on a specific `pos_label` (which is right now a constant)
* Our reductions create signed weights and they rely on the fact that the class encoding is the same between the meta-estimator (our wrapper) and base estimator.

For now, I suggest that we enforce/validate that y is valued 0/1 and treat this as a feature request rather than a bug.

**Comment by romanlutz** [comment]
Everything you're suggesting @MiroDudik is reasonable and can be put in a new issue as a feature-request. However, this is most certainly a bug if you check my example in the very first post as it changed the labels.

**Comment by riedgar-ms** [comment]
If I've understood correctly now, wouldn't requiring y to be 0/1 'fix' the bug (admittedly by rejecting otherwise valid input)?

**Comment by romanlutz** [comment]
Yes, but is that a reasonable restriction?

**Comment by adrinjalali** [comment]
Is the fix I suggest not what you'd like to see @riedgar-ms ?

**Comment by riedgar-ms** [comment]
I'm sure that a full solution would look something like that. It's more a case of how much of this we treat as a bug and how much we treat as a feature request. And I should be careful that I don't generate more debate about this than the energy required for a full fix ;-)

**Comment by MiroDudik** [comment]
How about this:
* let's raise an error if `y` has other values than 0/1; we could possibly mention that when transforming `y`, the user should pay attention to how the fairness constraints treat different values of `y`
* if `y` has only a single value and that value is either 0 or 1, we should proceed, but we should give a warning saying that we assume that the underlying `y` has binary encoding 0/1

**Comment by romanlutz** [comment]
@riedgar-ms 's PR addressed this for the short-term which addresses my originally raised concern. The long-term solution will be relevant with multi-class classification so we can think about it then.

**Comment by kalikhademi** [comment]
Is there any update on multi-class classification with fairlearn?

**Comment by romanlutz** [comment]
@kalikhademi I don't have updates since there's no ongoing work. If you'd like to submit a feature request with some context on what you're looking for please open a new issue. Thanks!
